src/threads.rs

```diff
@@ -0,0 +1,99 @@
+use std::cell::RefCell;
+use std::slice::IterMut;
+use std::sync::mpsc::{channel, Receiver, Sender};
+use std::thread::{self, JoinHandle};
+
+use crate::stats::RawSample;
+
+#[derive(Default)]
+pub(crate) struct AuxiliaryThreads {
+    inner: RefCell<Vec<BencherThread>>,
+}
+
+impl AuxiliaryThreads {
+    pub fn with_threads<ThreadFn, Scope, R>(
+        &self,
+        thread_count: usize,
+        task_fn: ThreadFn,
+        scope: Scope,
+    ) -> R
+    where
+        ThreadFn: Fn() -> RawSample,
+        Scope: FnOnce(Results) -> R,
+    {
+        let mut threads = self.inner.borrow_mut();
+        if threads.len() < thread_count {
+            threads.resize_with(thread_count, BencherThread::new)
+        }
+        let threads = &mut threads[..thread_count];
+
+        // SAFETY: We wait for all child task to finish executing their copy of the
+        // `task_fn` within `Results::drop`. Therefore we are not holding on to the `task_fn`
+        // after returning from this function / scope.
+        let dyn_task_fn: &dyn Fn() -> RawSample = &task_fn;
+        let dyn_task_fn: Task = unsafe { std::mem::transmute(dyn_task_fn) };
+
+        for thread in threads.iter() {
+            thread.send_workload.as_ref().unwrap().send(dyn_task_fn).unwrap();
+        }
+
+        scope(Results { threads: threads.iter_mut() })
+    }
+}
+
+pub(crate) struct Results<'t> {
+    threads: IterMut<'t, BencherThread>,
+}
+
+impl Iterator for Results<'_> {
+    type Item = RawSample;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let thread = self.threads.next()?;
+        Some(thread.expect_result())
+    }
+}
+
+impl Drop for Results<'_> {
+    fn drop(&mut self) {
+        for _result in self {}
+    }
+}
+
+type Task = &'static (dyn Fn() -> RawSample + Send + Sync);
```

Comment:
Since this is only used once, I would rather remove the type alias. Also, the original `Fn` is only `Sync`, not `Send`.

Reply:
its actually used twice, because the channel requires an explicit type annotation for whatever reason.